// flot-charting-plugin/dist/es5/jquery.flot.charting.js
(function (global) {

function CBuffer() {
	// handle cases where "new" keyword wasn't used
	if (!(this instanceof CBuffer)) {
		// multiple conditions need to be checked to properly emulate Array
		if (arguments.length > 1 || typeof arguments[0] !== 'number') {
			return CBuffer.apply(new CBuffer(arguments.length), arguments);
		} else {
			return new CBuffer(arguments[0]);
		}
	}
	// if no arguments, then nothing needs to be set
	if (arguments.length === 0)
	throw new Error('Missing Argument: You must pass a valid buffer length');
	// this is the same in either scenario
	this.size = this.start = 0;
	// set to callback fn if data is about to be overwritten
	this.overflow = null;
	// emulate Array based on passed arguments
	if (arguments.length > 1 || typeof arguments[0] !== 'number') {
		this.data = new Array(arguments.length);
		this.end = (this.length = arguments.length) - 1;
		this.push.apply(this, arguments);
	} else {
		this.data = new Array(arguments[0]);
		this.end = (this.length = arguments[0]) - 1;
	}
	// need to `return this` so `return CBuffer.apply` works
	return this;
}

function defaultComparitor(a, b) {
	return a == b ? 0 : a > b ? 1 : -1;
}

CBuffer.prototype = {
	// properly set constructor
	constructor : CBuffer,

	/* mutator methods */
	// pop last item
	pop : function () {
		var item;
		if (this.size === 0) return;
		item = this.data[this.end];
		// remove the reference to the object so it can be garbage collected
		delete this.data[this.end];
		this.end = (this.end - 1 + this.length) % this.length;
		this.size--;
		return item;
	},
	// push item to the end
	push : function (item) {
		// push items to the end, wrapping and erasing existing items
		this.data[(this.end + 1) % this.length] = item;
		// recalculate size
		if (this.size < this.length) {
			this.size ++;
		}
		// recalculate end
		this.end = (this.end + 1) % this.length;
		// recalculate start
		this.start = (this.length + this.end - this.size + 1) % this.length;
		// return number current number of items in CBuffer
		return this.size;
	},
	// reverse order of the buffer
	reverse : function () {
		var i = 0,
			tmp;
		for (; i < ~~(this.size / 2); i++) {
			tmp = this.data[(this.start + i) % this.length];
			this.data[(this.start + i) % this.length] = this.data[(this.start + (this.size - i - 1)) % this.length];
			this.data[(this.start + (this.size - i - 1)) % this.length] = tmp;
		}
		return this;
	},
	// rotate buffer to the left by cntr, or by 1
	rotateLeft : function (cntr) {
		if (typeof cntr === 'undefined') cntr = 1;
		if (typeof cntr !== 'number') throw new Error("Argument must be a number");
		while (--cntr >= 0) {
			this.push(this.shift());
		}
		return this;
	},
	// rotate buffer to the right by cntr, or by 1
	rotateRight : function (cntr) {
		if (typeof cntr === 'undefined') cntr = 1;
		if (typeof cntr !== 'number') throw new Error("Argument must be a number");
		while (--cntr >= 0) {
			this.unshift(this.pop());
		}
		return this;
	},
	// remove and return first item
	shift : function () {
		var item;
		// check if there are any items in CBuff
		if (this.size === 0) return;
		// store first item for return
		item = this.data[this.start];
		// recalculate start of CBuffer
		this.start = (this.start + 1) % this.length;
		// decrement size
		this.size--;
		return item;
	},
	// sort items
	sort : function (fn) {
		this.data.sort(fn || defaultComparitor);
		this.start = 0;
		this.end = this.size - 1;
		return this;
	},
	// add item to beginning of buffer
	unshift : function () {
		var i = 0;
		// check if overflow is set, and if data is about to be overwritten
		if (this.overflow && this.size + arguments.length > this.length) {
			// call overflow function and send data that's about to be overwritten
			for (; i < this.size + arguments.length - this.length; i++) {
				this.overflow(this.data[this.end - (i % this.length)], this);
			}
		}
		for (i = 0; i < arguments.length; i++) {
			this.data[(this.length + this.start - (i % this.length) - 1) % this.length] = arguments[i];
		}
		if (this.length - this.size - i < 0) {
			this.end += this.length - this.size - i;
			if (this.end < 0) this.end = this.length + (this.end % this.length);
		}
		if (this.size < this.length) {
			if (this.size + i > this.length) this.size = this.length;
			else this.size += i;
		}
		this.start -= arguments.length;
		if (this.start < 0) this.start = this.length + (this.start % this.length);
		return this.size;
	},

	/* accessor methods */
	// return index of first matched element
	indexOf : function (arg, idx) {
		if (!idx) idx = 0;
		for (; idx < this.size; idx++) {
			if (this.data[(this.start + idx) % this.length] === arg) return idx;
		}
		return -1;
	},
	// return last index of the first match
	lastIndexOf : function (arg, idx) {
		if (!idx) idx = this.size - 1;
		for (; idx >= 0; idx--) {
			if (this.data[(this.start + idx) % this.length] === arg) return idx;
		}
		return -1;
	},

	// return the index an item would be inserted to if this
	// is a sorted circular buffer
	sortedIndex : function(value, comparitor, context) {
		comparitor = comparitor || defaultComparitor;
		var low = this.start,
			high = this.size - 1;

		// Tricky part is finding if its before or after the pivot
		// we can get this info by checking if the target is less than
		// the last item. After that it's just a typical binary search.
		if (low && comparitor.call(context, value, this.data[high]) > 0) {
			low = 0, high = this.end;
		}

		while (low < high) {
		  var mid = (low + high) >>> 1;
		  if (comparitor.call(context, value, this.data[mid]) > 0) low = mid + 1;
		  else high = mid;
		}
		// http://stackoverflow.com/a/18618273/1517919
		return (((low - this.start) % this.size) + this.size) % this.size;
	},

	/* iteration methods */
	// check every item in the array against a test
	every : function (callback, context) {
		var i = 0;
		for (; i < this.size; i++) {
			if (!callback.call(context, this.data[(this.start + i) % this.length], i, this))
				return false;
		}
		return true;
	},
	// loop through each item in buffer
	// TODO: figure out how to emulate Array use better
	forEach : function (callback, context) {
		var i = 0;
		for (; i < this.size; i++) {
			callback.call(context, this.data[(this.start + i) % this.length], i, this);
		}
	},
	// check items agains test until one returns true
	// TODO: figure out how to emuldate Array use better
	some : function (callback, context) {
		var i = 0;
		for (; i < this.size; i++) {
			if (callback.call(context, this.data[(this.start + i) % this.length], i, this))
				return true;
		}
		return false;
	},
	// calculate the average value of a circular buffer
	avg : function () {
		return this.size == 0 ? 0 : (this.sum() / this.size);
	},
	// loop through each item in buffer and calculate sum
	sum : function () {
		var index = this.size;
		var s = 0;
		while (index--) s += this.data[index];
		return s;
	},
	// loop through each item in buffer and calculate median
	median : function () {
		if (this.size === 0)
			return 0;
		var values = this.slice().sort(defaultComparitor);
		var half = Math.floor(values.length / 2);
		if(values.length % 2)
			return values[half];
		else
			return (values[half-1] + values[half]) / 2.0;
	},
	/* utility methods */
	// reset pointers to buffer with zero items
	// note: this will not remove values in cbuffer, so if for security values
	//       need to be overwritten, run `.fill(null).empty()`
	empty : function () {
		var i = 0;
		this.size = this.start = 0;
		this.end = this.length - 1;
		return this;
	},
	// fill all places with passed value or function
	fill : function (arg) {
		var i = 0;
		if (typeof arg === 'function') {
			while(this.data[i] = arg(), ++i < this.length);
		} else {
			while(this.data[i] = arg, ++i < this.length);
		}
		// reposition start/end
		this.start = 0;
		this.end = this.length - 1;
		this.size = this.length;
		return this;
	},
	// return first item in buffer
	first : function () {
		return this.data[this.start];
	},
	// return last item in buffer
	last : function () {
		return this.data[this.end];
	},
	// return specific index in buffer
	get : function (arg) {
		return this.data[(this.start + arg) % this.length];
	},
	isFull : function (arg) {
		return this.length === this.size;
	},
	// set value at specified index
	set : function (idx, arg) {
		return this.data[(this.start + idx) % this.length] = arg;
	},
	// return clean array of values
	toArray : function () {
		return this.slice();
	},
	// slice the buffer to an array
	slice : function (start, end) {
		var length = this.size;

		start = +start || 0;

		if (start < 0) {
			if (start >= end)
				return [];
			start = (-start > length) ? 0 : length + start;
		}

		if (end == null || end > length)
			end = length;
		else if (end < 0)
			end += length;
		else
			end = +end || 0;

		length = start < end ? end - start : 0;

		var result = Array(length);
		for (var index = 0; index < length; index++) {
			result[index] = this.data[(this.start + start + index) % this.length];
		}
		return result;
	}
};

if (typeof module === 'object' && module.exports) module.exports = CBuffer;
else global.CBuffer = CBuffer;

}(window));

/* Segment tree data structure for fast min/max queries

Copyright (c) 2007-2015 National Instruments
Licensed under the MIT license.
*/
/*globals CBuffer, module*/

(function (global) {
    'use strict';

    /* a TreeNode object keeps information about min and max values in the subtree below it*/
    var SegmentTreeNode = function () {
        this.init();
    };

    /* a SegmentTree holds the data used for accelerated query of the history buffer*/
    var SegmentTree = function (hb, cbuffer) {
        this.historyBuffer = hb;
        this.cbuffer = cbuffer;
        this.tree = this.buildEmptySegmentTree();
    };

    SegmentTreeNode.prototype.init = function () {
        this.maxIndex = 0;
        this.minIndex = 0;
        this.max = Math.Nan;
        this.min = Math.Nan;
    };

    /* a tree level is a heap of tree nodes at a certain depth in the tree*/
    var SegmentTreeLevel = function (historyBuffer, level) {
        this.level = level;
        this.step = Math.pow(historyBuffer.branchFactor, level);
        this.capacity = Math.ceil(historyBuffer.capacity / (Math.pow(historyBuffer.branchFactor, level))) + 1;
        this.startIndex = 0;
        this.nodes = new CBuffer(this.capacity);
    };

    /* rotate the nodes in the TreeLevel to the left.*/
    SegmentTreeLevel.prototype.rotate = function () {
        this.startIndex += this.step;

        var oldestNode = this.nodes.shift(); //reuse the tree nodes to reduce GC
        oldestNode.init();
        this.nodes.push(oldestNode);
    };

    /* reinitialize the nodes in the TreeLevel.*/
    SegmentTreeLevel.prototype.init = function (startIndex) {
        this.startIndex = startIndex;

        this.nodes.forEach(function (node) {
            node.init();
        });
    };

    /*get the nth element in the buffer*/
    SegmentTree.prototype.get = function (index) {
        index -= this.historyBuffer.startIndex();
        return this.cbuffer.get(index);
    };

    /* get the tree node at the specified level that keeps the information for the specified index*/
    SegmentTree.prototype.getTreeNode = function (level, index) {
        var treeLevel = this.tree.levels[level];
        var levelStep = treeLevel.step;
        var levelIndex = Math.floor((index - treeLevel.startIndex) / levelStep);

        if ((levelIndex < 0) || (levelIndex >= treeLevel.capacity)) {
            return null;
        }

        var node = treeLevel.nodes.get(levelIndex);

        return node;
    };

    /* builds an empty acceleration tree*/
    SegmentTree.prototype.buildEmptySegmentTree = function () {
        var hb = this.historyBuffer;
        var depth = Math.ceil(Math.log(hb.capacity) / Math.log(hb.branchFactor)) - 1;
        if (depth < 1) {
            depth = 1;
        }

        var tree = {
            depth: depth,
            levels: []
        };

        for (var i = 0; i < depth; i++) {
            var tLevel = new SegmentTreeLevel(hb, i + 1);
            tree.levels.push(tLevel);
            for (var j = 0; j < tLevel.capacity; j++) {
                var node = new SegmentTreeNode();
                tLevel.nodes.push(node);
            }
        }

        return tree;
    };

    /*
     * Populate the upper levels of the tree, starting at the startingFromIndex.
     * All the tree levels should be already shifted as necessary before calling this function.
     */
    SegmentTree.prototype.populateTreeLevel = function (startingFrom, level) {
        var hb = this.historyBuffer;
        var cbuffer = this.cbuffer;
        var startIndex = hb.startIndex(); // cache it
        var lastIndex = hb.lastIndex(); // cache it
        var currentCount = 0;
        var i = 0;
        var firstSample = true;
        var node, max, maxIndex, min, minIndex;

        var minusOneLevel = {
            step: 1,
            startIndex: startIndex
        };

        var baseLevel = (level === 0) ? minusOneLevel : this.tree.levels[level - 1];
        var currentLevel = this.tree.levels[level];

        /* align starting from to a node in the base level boundary*/
        startingFrom = floorInBase(startingFrom, currentLevel.step);

        if (baseLevel.startIndex > startingFrom) {
            startingFrom = baseLevel.startIndex;
            currentCount = (startingFrom / baseLevel.step) % hb.branchFactor;
        }

        for (i = startingFrom; i < lastIndex; i += baseLevel.step) {
            if (level === 0) {
                var val = cbuffer.get(i - startIndex);

                if (firstSample) {
                    max = val;
                    maxIndex = i;
                    min = val;
                    minIndex = i;

                    firstSample = false;
                } else {
                    if (val > max) {
                        max = val;
                        maxIndex = i;
                    }

                    if (val < min) {
                        min = val;
                        minIndex = i;
                    }
                }
            } else {
                var cNode = this.getTreeNode(level - 1, i);

                if (firstSample) {
                    max = cNode.max;
                    maxIndex = cNode.maxIndex;
                    min = cNode.min;
                    minIndex = cNode.minIndex;
                    firstSample = false;
                } else {
                    if (cNode.max > max) {
                        max = cNode.max;
                        maxIndex = cNode.maxIndex;
                    }

                    if (cNode.min < min) {
                        min = cNode.min;
                        minIndex = cNode.minIndex;
                    }
                }
            }

            currentCount++;

            if (currentCount === hb.branchFactor) {
                currentCount = 0;
                firstSample = true;
                node = this.getTreeNode(level, i);

                node.max = max;
                node.maxIndex = maxIndex;
                node.min = min;
                node.minIndex = minIndex;
            }
        }

        if (currentCount !== 0) {
            node = this.getTreeNode(level, i);

            node.max = max;
            node.maxIndex = maxIndex;
            node.min = min;
            node.minIndex = minIndex;
        }
    };

    /* Rotate the history buffer to the left, updating the leftmost nodes in the tree with the new mins and maxes*/
    SegmentTree.prototype.rotateTreeLevel = function (level) {
        var hb = this.historyBuffer;

        var startingIndex = hb.startIndex();
        var treeLevel = this.tree.levels[level];

        var alignedStartIndex = floorInBase(startingIndex, treeLevel.step);

        if (alignedStartIndex - treeLevel.startIndex > hb.capacity) {
            treeLevel.init(alignedStartIndex);
        } else {
            while (treeLevel.startIndex < alignedStartIndex) {
                treeLevel.rotate();
            }
        }

        /* update the first node in the level */
        if (startingIndex !== alignedStartIndex) {
            var minmax = {
                minIndex: startingIndex,
                min: this.get(startingIndex),
                maxIndex: startingIndex,
                max: this.get(startingIndex)
            };

            var i;
            var firstNode = treeLevel.nodes.get(0);

            if (level === 0) {
                for (i = startingIndex; i < (alignedStartIndex + hb.branchFactor); i++) {
                    updateMinMaxFromValue(i, this.get(i), minmax);
                }
            } else {
                for (i = startingIndex; i < (alignedStartIndex + treeLevel.step); i += treeLevel.step / hb.branchFactor) {
                    updateMinMaxFromNode(this.getTreeNode(level - 1, i), minmax);
                }
            }

            firstNode.minIndex = minmax.minIndex;
            firstNode.min = minmax.min;
            firstNode.maxIndex = minmax.maxIndex;
            firstNode.max = minmax.max;
        }
    };

    SegmentTree.prototype.updateSegmentTree = function () {
        var level;

        for (level = 0; level < this.tree.depth; level++) {
            this.rotateTreeLevel(level);
        }

        for (level = 0; level < this.tree.depth; level++) {
            this.populateTreeLevel(this.historyBuffer.lastUpdatedIndex, level);
        }

    };

    SegmentTree.prototype.readMinMax = function (start, end, minmax) {
        var intervalSize = end - start;
        var hb = this.historyBuffer;
        var startIndex = hb.startIndex();
        var cbuffer = this.cbuffer;

        var i;

        var level = Math.floor(Math.log(intervalSize) / Math.log(hb.branchFactor));

        if (level <= 0) {
            for (i = start; i < end; i++) {
                updateMinMaxFromValue(i, cbuffer.get(i - startIndex), minmax);
            }

            return minmax;
        }

        if (level > this.tree.levels.length) {
            level =  this.tree.levels.length;
        }

        var step = Math.pow(hb.branchFactor, level);
        var truncatedStart = Math.ceil(start / step) * step;
        var truncatedEnd = floorInBase(end, step);

        if (start !== truncatedStart) {
            this.readMinMax(start, truncatedStart, minmax);
        }

        var truncatedBufferStart = floorInBase(startIndex, step);
        var begin = (truncatedStart - truncatedBufferStart) / step;
        var finish = (truncatedEnd - truncatedBufferStart) / step;

        for (i = begin; i < finish; i++) {
            updateMinMaxFromNode(this.tree.levels[level - 1].nodes.get(i), minmax);
        }

        if (end !== truncatedEnd) {
            this.readMinMax(truncatedEnd, end, minmax);
        }

        return minmax;
    };

    /* get a decimated series, starting at the start sample, ending at the end sample with a provided step */
    SegmentTree.prototype.query = function (start, end, step) {
        var i;
        var hb = this.historyBuffer;
        start = Math.floor(start);
        end = Math.ceil(end);

        var data = [];

        var firstIndex = hb.startIndex();
        var lastIndex = hb.lastIndex();

        if (start < firstIndex) {
            start = firstIndex;
        }

        if (start > lastIndex) {
            start = lastIndex;
        }

        if (end < firstIndex) {
            end = firstIndex - 1;
        }

        if (end > lastIndex-1) {
            end = lastIndex-1;
        }

        if (step < 4) { // for small steps it is more efficient to bypass the segment tree. TODO: benchmark this
            for (i = start; i <= end; i++) {
                data.push(i);
                data.push(this.get(i));
            }
        } else {
            var minmax = new SegmentTreeNode();

            var maxIndex, minIndex;
            for (i = start; i <= end; i = Math.round( i + step)) {
                var partialQueryEnd = Math.min(end + 1, Math.round(i + step));
                minmax.max = Number.NEGATIVE_INFINITY;
                minmax.min = Number.POSITIVE_INFINITY;
                minmax.minIndex = 0;
                minmax.maxIndex = 0;

                minmax = this.readMinMax(i, partialQueryEnd, minmax);
                maxIndex = minmax.maxIndex;
                minIndex = minmax.minIndex;
                if (minIndex === maxIndex) {
                    data.push(minIndex);
                    data.push(minmax.min);
                } else if (minIndex < maxIndex) {
                    data.push(minIndex);
                    data.push(minmax.min);
                    data.push(maxIndex);
                    data.push(minmax.max);
                } else {
                    data.push(maxIndex);
                    data.push(minmax.max);
                    data.push(minIndex);
                    data.push(minmax.min);
                }
            }
        }

        return data;
    };

    function updateMinMaxFromValue(index, value, minmax) {
        if (value < minmax.min) {
            minmax.min = value;
            minmax.minIndex = index;
        }

        if (value > minmax.max) {
            minmax.max = value;
            minmax.maxIndex = index;
        }
    }

    function updateMinMaxFromNode(node, minmax) {
        if (node.min < minmax.min) {
            minmax.min = node.min;
            minmax.minIndex = node.minIndex;
        }

        if (node.max > minmax.max) {
            minmax.max = node.max;
            minmax.maxIndex = node.maxIndex;
        }
    }

    // round to nearby lower multiple of base
    function floorInBase(n, base) {
        return base * Math.floor(n / base);
    }

    if (typeof module === 'object' && module.exports) {
        module.exports = SegmentTree;
    } else {
        global.SegmentTree = SegmentTree;
    }
})(window);

/* history buffer data structure for charting.

Copyright (c) 2007-2015 National Instruments
Licensed under the MIT license.
*/
/*globals CBuffer, SegmentTree, module*/

/**
# HistoryBuffer

> A historyBuffer is a data structure that enables efficient charting operations
on a sliding window of data points.

In the case of large data buffers it is inefficient to draw every point of the
chart. Doing this results in many almost vertical lines drawn over the same
stripe of pixels over and over again. Drawing a line on a canvas is an expensive
operation that must be avoided if possible.

One method of avoiding the repeated drawing is to reduce the amount of data points
we draw on the chart by sub-sampling the data, also called decimation.

There are many ways to decimate the data; the one this history buffer implements
is to divide data into "1 pixel wide buckets" and then for each bucket select the
maximum and minimum as subsamples. This method results in a drawing that looks
visually similar with the one in which all samples are drawn.

The history buffer is a circular buffer holding the chart data accompanied by an
acceleration structure - a segment tree of min/max values.

The segment tree is only enabled for big history buffers.

Example:
```javascript
var hb1 = new HistoryBuffer(1024);

// in a history buffer with width 1 we can push scalars
hb1.push(1);
hb1.push(2);

console.log(hb1.toArray()); //[1, 2]

// as well as 1 elements arrays
hb1.push([3]);
hb1.push([4]);

console.log(hb1.toArray()); //[1, 2, 3, 4]

// or append an array
hb1.appendArray([5, 6]);

console.log(hb1.toArray()); //[1, 2, 3, 4, 5, 6]
```

The history buffer is able to store multiple "parallel" data sets. Example:

```javascript
var hb2 = new HistoryBuffer(1024, 2);

// in a history buffer with width > 1 we can only push arrays
hb2.push([1, 11]);
hb2.push([2, 12]);
hb2.push([3, 13]);

console.log(hb2.toArray()); //[[1, 11], [2, 12], [3, 13]]

// or append an array of arrays
hb2.appendArray([[4, 14], [5, 15], [6, 16]]);

console.log(hb2.toArray()); //[[1, 11], [2, 12], [3, 13], [4, 14], [5, 15], [6, 16]]
```

Operations accelerated by a historyBuffer
-----------------------------------------
The common charting operations performed on a history buffer are

* inserting elements at the head
* inserting m elements at the head
* deleting elements at the tail
* deleting m elements at the tail
* compute min/max on a range
* query for a "visually interesting" data subsample on a range

*/

/** ## HistoryBuffer methods*/
(function (global) {
    'use strict';

    /* The branching factor determines how many samples are decimated in a tree node.
     * It affects the performance and the overhead of the tree.
     */
    var defaultBranchFactor = 32; // 32 for now. TODO tune the branching factor.

    /** **HistoryBuffer(capacity, width)** - the History buffer constructor creates
    a new history buffer with the specified capacity (default: 1024) and width (default: 1)*/
    var HistoryBufferNumeric = function (capacity, width) {
        this.capacity = capacity || 1024;
        this.width = width || 1;
        this.lastUpdatedIndex = 0;
        this.firstUpdatedIndex = 0;
        this.branchFactor = defaultBranchFactor;

        this.buffers = []; // circular buffers for data
        this.trees = []; // segment trees

        for (var i = 0; i < this.width; i++) {
            this.buffers.push(new CBuffer(capacity));
            this.trees.push(new SegmentTree(this, this.buffers[i]));
        }

        this.buffer = this.buffers[0];
        this.tree = this.trees[0];

        this.count = 0;
        this.callbacks = new Map();
        this.changed = false;
    };

    HistoryBufferNumeric.prototype.callOnChange = function (args) {
        this.callbacks.forEach(callback => {
            callback(args);
        });
    };

    HistoryBufferNumeric.prototype.setBranchingFactor = function (b) {
        this.branchFactor = b;

        this.rebuildSegmentTrees();
    };

    HistoryBufferNumeric.prototype.getDefaultBranchingFactor = function () {
        return defaultBranchFactor;
    };

    HistoryBufferNumeric.prototype.rebuildSegmentTrees = function () {
        this.trees = []; // new segment trees

        for (var i = 0; i < this.width; i++) {
            this.trees.push(new SegmentTree(this, this.buffers[i]));
        }

        this.tree = this.trees[0];

        this.firstUpdatedIndex = this.startIndex();
        this.lastUpdatedIndex = this.firstUpdatedIndex;

        this.updateSegmentTrees();
    };

    /** **clear()** - clears the history buffer */
    HistoryBufferNumeric.prototype.clear = function () {
        for (var i = 0; i < this.width; i++) {
            this.buffers[i].empty();
        }

        this.count = 0; // todo fire changes and upate lastindex, startindex
        this.rebuildSegmentTrees();
        this.changed = true;
        if (this.callOnChange) {
            this.callOnChange();
        }
    };

    /** **setCapacity(newCapacity)** changes the capacity of the History Buffer and clears all the data inside it */
    HistoryBufferNumeric.prototype.setCapacity = function (newCapacity) {
        if (newCapacity !== this.capacity) {
            this.capacity = newCapacity;
            this.buffers = []; // circular buffers for data

            for (var i = 0; i < this.width; i++) {
                this.buffers.push(new CBuffer(newCapacity));
            }

            this.buffer = this.buffers[0];
            this.count = 0; // todo fire changes and upate lastindex, startindex
            this.rebuildSegmentTrees();
            this.changed = true;
            if (this.callOnChange) {
                this.callOnChange();
            }
        }
    };

    /** **setWidth(newWidth)** - changes the width of the History Buffer and clears
    all the data inside it */
    HistoryBufferNumeric.prototype.setWidth = function (newWidth) {
        if (newWidth !== this.width) {
            this.width = newWidth;
            this.buffers = []; // clear the circular buffers for data. TODO reuse the buffers

            for (var i = 0; i < this.width; i++) {
                this.buffers.push(new CBuffer(this.capacity));
            }

            this.buffer = this.buffers[0];
            this.count = 0; // todo fire changes and upate lastindex, startindex
            this.rebuildSegmentTrees();
            this.changed = true;
            if (this.callOnChange) {
                this.callOnChange();
            }
        }
    };

    /* returns the current width of History Buffer */ 
    HistoryBufferNumeric.prototype.getWidth = function () {
        return this.width;
    };

    /* store an element in the history buffer, don't update stats */
    HistoryBufferNumeric.prototype.pushNoStatsUpdate = function (item) {
        if (typeof item === 'number' && this.width === 1) {
            this.buffer.push(item);
        } else {
            if (Array.isArray(item) && item.length === this.width) {
                for (var i = 0; i < this.width; i++) {
                    this.buffers[i].push(item[i]);
                }
            }
        }
    };

    /** **push(item)** - adds an element to the history buffer */
    HistoryBufferNumeric.prototype.push = function (item) {
        this.pushNoStatsUpdate(item);
        this.count++;

        this.changed = true;
        if (this.callOnChange) {
            this.callOnChange();
        }

    };

    /** **startIndex()** - returns the index of the oldest element in the buffer*/
    HistoryBufferNumeric.prototype.startIndex = function () {
        return Math.max(0, this.count - this.capacity);
    };

    /** **lastIndex()** - returns the index of the newest element in the buffer*/
    HistoryBufferNumeric.prototype.lastIndex = function () {
        return this.startIndex() + this.buffer.size;
    };

    /** **get(n)** - returns the nth element in the buffer*/
    HistoryBufferNumeric.prototype.get = function (index) {
        index -= this.startIndex();
        if (this.width === 1) {
            return this.buffer.get(index);
        } else {
            var res = [];

            for (var i = 0; i < this.width; i++) {
                res.push(this.buffers[i].get(index));
            }

            return res;
        }
    };

    /** **appendArray(arr)** - appends an array of elements to the buffer*/
    HistoryBufferNumeric.prototype.appendArray = function (arr) {
        for (var i = 0; i < arr.length; i++) {
            this.pushNoStatsUpdate(arr[i]);
        }

        this.count += arr.length;

        this.changed = true;
        if (this.callOnChange) {
            this.callOnChange();
        }
    };

    /** **toArray()** - returns the content of the history buffer as an array */
    HistoryBufferNumeric.prototype.toArray = function () {
        if (this.width === 1) {
            return this.buffer.toArray();
        } else {
            var start = this.startIndex(),
                last = this.lastIndex(),
                res = [];
            for (var i = start; i < last; i++) {
                res.push(this.get(i));
            }

            return res;
        }
    };

    /* update the segment tree with the newly added values*/
    HistoryBufferNumeric.prototype.updateSegmentTrees = function () {
        var buffer = this.buffer;

        this.trees.forEach(function (tree) {
            tree.updateSegmentTree();
        });

        this.firstUpdatedIndex = this.startIndex();
        this.lastUpdatedIndex = this.firstUpdatedIndex + buffer.size;
    };

    /** **toDataSeries()** - returns the content of the history buffer into a
    flot data series*/
    HistoryBufferNumeric.prototype.toDataSeries = function (index) {
        var buffer = this.buffer;

        var data = [];

        var start = this.startIndex();

        for (var i = 0; i < buffer.size; i++) {
            data.push([i + start, this.buffers[index || 0].get(i)]);
        }

        return data;
    };

    HistoryBufferNumeric.prototype.registerOnChange = function (key, f) {
        if (this.callbacks.has(key)) {
            console.log('Callback with the given key has already been registered.');
            return;
        }

        this.callbacks.set(key, f);
    };

    HistoryBufferNumeric.prototype.deregisterOnChange = function (key) {
        this.callbacks.delete(key);
    };

    /** **query(start, end, step, index)** - decimates the data set at the
    provided *index*, starting at the start sample, ending at the end sample
    with the provided step */
    HistoryBufferNumeric.prototype.query = function (start, end, step, index) {
        if (index === undefined) {
            index = 0;
        }

        if (this.changed) {
            this.updateSegmentTrees();
            this.changed = false;
        }

        return this.trees[index].query(start, end, step);
    };

    /** **rangeX( index)** - returns the range of the data in the buffer*/
    HistoryBufferNumeric.prototype.rangeX = function (index) {
        var start = this.startIndex(),
            end = this.lastIndex()-1;

        if (end === start - 1) {
            return {};
        }

        return { xmin: start,
                 xmax: end,
                 deltamin: 1
               };
    };

    /** **rangeY(start, end, index)** - returns the range of the data
    in a given interval of the buffer*/
    HistoryBufferNumeric.prototype.rangeY = function (start, end, index) {
        if (start === null || start === undefined){
            start = this.startIndex();
        }
        if (end === null || end === undefined){
            end = this.lastIndex()-1;
        }
        if (index === null || index === undefined) {
            index = 0;
        }

        if (this.changed) {
            this.updateSegmentTrees();
            this.changed = false;
        }

        var data = this.query(start, end, end - start, index),
            dataLength = data.length;

        if (dataLength > 0) {
            var res = {
                ymin: Infinity,
                ymax: -Infinity
            };

            for (var i = 0; i < dataLength; i += 2) {
                if (data[i + 1] == null || data[i + 1] === Infinity || data[i + 1] === -Infinity || isNaN(data[i + 1])) {
                    continue;
                }
                res.ymin = Math.min(res.ymin, data[i + 1]);
                res.ymax = Math.max(res.ymax, data[i + 1]);
            }

            return res;
        }
        return { };
    };

    HistoryBufferNumeric.prototype.toJSON = function() {
        var serializedHb = {
            data: [],
            width: this.width,
            capacity: this.capacity,
            valueType: 'HistoryBuffer',
            startIndex: this.startIndex(),
            count: this.count
        };

        if(this.width === 1) {
            serializedHb['data'] = this.buffer.toArray();
        } else {
            for (var i = 0; i < this.width; i++) {
                    serializedHb['data'].push(this.buffers[i].toArray());
            }
        }

        return serializedHb;
    };

    if (typeof module === 'object' && module.exports) {
        module.exports = HistoryBufferNumeric;
    } else {
        global.HistoryBufferNumeric = HistoryBufferNumeric;
    }
})(window);

/* history buffer data structure for charting.

Copyright (c) 2007-2015 National Instruments
Licensed under the MIT license.
*/
/*globals CBuffer, SegmentTree, module*/

/**
# HistoryBufferWaveform

> A historyBuffer is a data structure that enables efficient charting operations
on a sliding window of data points. HistoryBufferWaveform enables a charting of analog waveforms.

Example:
```javascript
var hb1 = new HistoryBufferWaveform(10);

aw = new NIAnalogWaveform({
    t0: 4,
    dt: 1,
    Y:[1, 2, 3]
});

aw1 = new NIAnalogWaveform({
    t0: 1,
    dt: 1,
    Y:[1, 2, 3]
});

// in an analog waveform history buffer with width 1 we can push analog waveforms
hb1.push(aw);
hb1.push(aw1);

console.log(hb1.toArray()); //[aw, aw2]

console.log(hb1.toDataSeries()); //[[4, 1], [5, 2], [6, 3], [null, null], [1, 1], [2, 2], [3, 3]]
```

*/

/** ## HistoryBufferWaveform methods*/
(function (global) {
    'use strict';

    /** **HistoryBufferWaveform(capacity, width)** - the History buffer constructor creates
    a new history buffer with the specified capacity (default: 1024) and width (default: 1)*/
    var HistoryBufferWaveform = function (capacity, width) {
        this.capacity = capacity || 1024;
        this.width = width || 1;

        this.buffers = []; // circular buffers for data

        for (var i = 0; i < this.width; i++) {
            this.buffers.push(new CBuffer(capacity));
        }

        this.buffer = this.buffers[0];

        this.count = 0;
        this.callbacks = new Map();
        this.changed = false;
    };

    HistoryBufferWaveform.prototype.__proto__ = HistoryBufferNumeric.prototype; // delegate to HistoryBuffer

    HistoryBufferWaveform.prototype.rebuildSegmentTrees = function () { // no segment tree is used for waveforms
    };

    /* store an element in the history buffer, don't update stats */
    HistoryBufferWaveform.prototype.pushNoStatsUpdate = function (item) {
        if (this.width === 1) {
            this.buffer.push(item);
        } else {
            if (Array.isArray(item) && item.length === this.width) {
                for (var i = 0; i < this.width; i++) {
                    this.buffers[i].push(item[i]);
                }
            }
        }
    };

    /* get the tree nodes at the specified level that keeps the information for the specified interval*/
    HistoryBufferWaveform.prototype.getTreeNodes = function (level, start, end) { // no segment tree is used for waveforms
    };

    /* update the segment tree with the newly added values*/
    HistoryBufferWaveform.prototype.updateSegmentTrees = function () {
    };

    function waveformInRange(aw, start, end) {
        if (aw.Y.length === 0) {
            return false;
        }

        var t0 = new NITimestamp(aw.t0);

        var waveformStart = t0.valueOf();
        var waveformEnd = t0.valueOf() + aw.Y.length * aw.dt;

        if (waveformStart < start && waveformEnd < start) {
            return false;
        }

        if (waveformStart > end && waveformEnd > end) {
            return false;
        }

        return true;
    }

    function waveformsSeparated(aw1, aw2) {
        if (aw1.Y.length == 1 || aw2.Y.length == 1) {
            return false;
        }

        var lastSampleOfFirstWaveform = (new NITimestamp(aw1.t0)).add(aw1.dt * (aw1.Y.length - 1)).valueOf();
        var firstSampleOfSecondWaveform = new NITimestamp(aw2.t0).valueOf();

        return firstSampleOfSecondWaveform - lastSampleOfFirstWaveform >= 2.0 * aw1.dt;
    }

    function appendWaveformToDecimateBuffer(aw, start, end, buffer) {
        var Y = aw.Y,
            TS = aw.t0,
            currentTS = new NITimestamp(TS),
            floatCurrentTS;

        for (var i=0; i < Y.length; i++) {
            floatCurrentTS = currentTS.valueOf();

            if (floatCurrentTS >= (start - aw.dt) && floatCurrentTS <= (end + aw.dt)) {
                buffer.push(floatCurrentTS);
                buffer.push(Y[i]);
            }
            currentTS.add(aw.dt);
        }
    }

    function appendWaveformToDataSeries(aw, buffer) {
        var Y = aw.Y,
            TS = aw.t0,
            currentTS = new NITimestamp(TS),
            floatCurrentTS;

        for (var i=0; i < Y.length; i++) {
            floatCurrentTS = currentTS.valueOf();
            buffer.push([floatCurrentTS, Y[i]]);
            currentTS.add(aw.dt);
        }
    }

    /** **query(start, end, step, index)** - decimates the data set at the
    provided *index*, starting at the start sample, ending at the end sample
    with the provided step */
    HistoryBufferWaveform.prototype.query = function (start, end, step, index) {
        if (index === undefined) {
            index = 0;
        }

        var result = [];
        var waveforms = this.buffers[index].toArray();
        var previousWaveform = waveforms[0];

        waveforms.forEach(function (waveform) {
            if (!waveformInRange(waveform, start, end)) {
                return;
            }

            if (waveformsSeparated(previousWaveform, waveform)) {
                // add a "gap" to separate the analog waveforms
                result.push(null);
                result.push(null);
            }

            previousWaveform = waveform;
            appendWaveformToDecimateBuffer(waveform, start, end, result);
        });

        return result;
    };

    /** **toDataSeries()** - returns the content of the history buffer into a
    flot data series*/
    HistoryBufferWaveform.prototype.toDataSeries = function (index) {
        if (index === undefined) {
            index = 0;
        }

        var result = [];
        var waveforms = this.buffers[index].toArray();
        var previousWaveform = waveforms[0];

        waveforms.forEach(function (waveform) {
            if (waveformsSeparated(previousWaveform, waveform)) {
                // add a "gap" to separate the analog waveforms
                result.push([null, null]);
            }

            previousWaveform = waveform;
            appendWaveformToDataSeries(waveform, result);
        });

        return result;
    };

    HistoryBufferWaveform.prototype.rangeX = function (index) {
        var minTS = Infinity,
            maxTS = -Infinity,
            deltamin = Infinity,
            t0, startTS, endTS;

        if (index === undefined) {
            index = 0;
        }

        var waveforms = this.buffers[index].toArray();

        if (waveforms.length === 0) {
            return {};
        }

        waveforms.forEach(function (aw) {
            t0 = new NITimestamp(aw.t0);
            startTS = t0.valueOf();
            endTS = (new NITimestamp(t0)).add(aw.dt * (aw.Y.length - 1)).valueOf();

            if (startTS < minTS) {
                minTS = startTS;
            }

            if (endTS > maxTS) {
                maxTS = endTS;
            }

            if (deltamin > Math.abs(aw.dt)) {
                deltamin = Math.abs(aw.dt);
            }
        });

        if (deltamin === 0) {
            deltamin = 1;
        }

        if (minTS > maxTS) {
            var temp = minTS;
            minTS = maxTS;
            maxTS = temp;
        }

        return {
            xmin: minTS,
            xmax: maxTS,
            deltamin: deltamin
        }
    };

    HistoryBufferWaveform.prototype.rangeY = function (start, end, index) {
        var minMax = {min : Infinity,
                      max : -Infinity}

        if (index === null || index === undefined) {
            index = 0;
        }

        var waveforms = this.buffers[index].toArray();

        if (waveforms.length === 0) {
            return {};
        }

        if (start === null || start === undefined){
            start = (new NITimestamp(waveforms[0].t0)).valueOf();
        }
        if (end === null || end === undefined){
            var aw = waveforms[waveforms.length - 1];
            end = (new NITimestamp(aw.t0)).add(aw.dt * aw.Y.length).valueOf();
        }

        waveforms.forEach(function (waveform) {
            updateMinMax(waveform, minMax, start, end);
        });

        return {
            ymin: minMax.min,
            ymax: minMax.max
        }

    }

    function updateMinMax(aw, minMax, start, end) {
        var startTS, endTS, t,
            Y = aw.Y,
            t0 = new NITimestamp(aw.t0);

        if (Y.length === 0) {
            return;
        }

        startTS = t0.valueOf();
        endTS = (new NITimestamp(t0)).add(aw.dt * aw.Y.length).valueOf();

        if (startTS > end || endTS < start) {
            return;
        }

        for (var i = 0; i < Y.length; i++) {
            t = (new NITimestamp(t0)).add(aw.dt * i).valueOf();
            if (t < start || t > end) {
                continue;
            }

            if (Y[i] == null || Y[i] === Infinity || Y[i] === -Infinity || isNaN(Y[i])) {
                continue;
            }

            if (Y[i] > minMax.max) {
                minMax.max = Y[i];
            }

            if (Y[i] < minMax.min) {
                minMax.min = Y[i];
            }
        }
    }


    HistoryBufferWaveform.prototype.toJSON = function() {
        var serializedHb = {
            data: this.toArray(),
            width: this.width,
            capacity: this.capacity,
            valueType: 'HistoryBuffer',
            startIndex: this.startIndex(),
            count: this.count
        };

        return serializedHb;
    };

    if (typeof module === 'object' && module.exports) {
        module.exports = HistoryBufferWaveform;
    } else {
        global.HistoryBufferWaveform = HistoryBufferWaveform;
    }
})(window);

/* history buffer data structure for charting, supporting numeric and waveform data

Copyright (c) 2007-2015 National Instruments
Licensed under the MIT license.
*/
/*globals CBuffer, SegmentTree, module*/

/**
# HistoryBuffer

> A historyBuffer is a data structure that enables efficient charting operations
on a sliding window of data points.

Example:
```javascript
var hb1 = new HistoryBuffer(1024);

// in a history buffer with width 1 we can push scalars
hb1.push(1);
hb1.push(2);

console.log(hb1.toArray()); //[1, 2]

// as well as 1 elements arrays
hb1.push([3]);
hb1.push([4]);

console.log(hb1.toArray()); //[1, 2, 3, 4]

// or append an array
hb1.appendArray([5, 6]);

console.log(hb1.toArray()); //[1, 2, 3, 4, 5, 6]
```

The history buffer is able to store multiple "parallel" data sets. Example:

```javascript
var hb2 = new HistoryBuffer(1024, 2);

// in a history buffer with width > 1 we can only push arrays
hb2.push([1, 11]);
hb2.push([2, 12]);
hb2.push([3, 13]);

console.log(hb2.toArray()); //[[1, 11], [2, 12], [3, 13]]

// or append an array of arrays
hb2.appendArray([[4, 14], [5, 15], [6, 16]]);

console.log(hb2.toArray()); //[[1, 11], [2, 12], [3, 13], [4, 14], [5, 15], [6, 16]]
```

Operations accelerated by a historyBuffer
-----------------------------------------
The common charting operations performed on a history buffer are

* inserting elements at the head
* inserting m elements at the head
* deleting elements at the tail
* deleting m elements at the tail
* compute min/max on a range
* query for a "visually interesting" data subsample on a range

*/

/** ## HistoryBuffer methods*/
(function (global) {
    'use strict';

    /** **HistoryBuffer(capacity, width)** - the History buffer constructor creates
    a new history buffer with the specified capacity (default: 1024) and width (default: 1)*/
    var HistoryBuffer = function (capacity, width) {
        this.hb = new HistoryBufferNumeric(capacity, width);
        this.hbType = 'numeric';
    };

    /** **setType(type)** - sets the type of the history buffer. Accepted values are 'numeric' and 'analogWaveform' */
    HistoryBuffer.prototype.setType = function (type) {
        if (type === this.hbType) {
            return;
        }

        if (type === 'numeric') {
            var callbacks = this.hb.callbacks;
            this.hb.callbacks = undefined;
            this.hb = new HistoryBufferNumeric(this.capacity, this.width);
            callbacks.forEach((callback, key) => {
                this.hb.registerOnChange(key, callback);
            });

            this.hbType = 'numeric';
            this.hb.callOnChange();
        }

        if (type === 'analogWaveform') {
            var callbacks = this.hb.callbacks;
            this.hb.callbacks = undefined;
            this.hb = new HistoryBufferWaveform(this.capacity, this.width);
            callbacks.forEach((callback, key) => {
                this.hb.registerOnChange(key, callback);
            });
            this.hbType = 'analogWaveform';
            this.hb.callOnChange();
        }
    };

    /** **clear()** - clears the history buffer */
    HistoryBuffer.prototype.clear = function () {
        this.hb.clear();
    };

    /** **setCapacity(newCapacity)** changes the capacity of the History Buffer and clears all the data inside it */
    HistoryBuffer.prototype.setCapacity = function (newCapacity) {
        this.hb.setCapacity(newCapacity);
    };

    /** **setWidth(newWidth)** - changes the width of the History Buffer and clears
    all the data inside it */
    HistoryBuffer.prototype.setWidth = function (newWidth) {
        this.hb.setWidth(newWidth);
    };

    /** **push(item)** - adds an element to the history buffer */
    HistoryBuffer.prototype.push = function (item) {
        this.hb.push(item);
    };

    /** **startIndex()** - returns the index of the oldest element in the buffer*/
    HistoryBuffer.prototype.startIndex = function () {
        return this.hb.startIndex();
    };

    /** **lastIndex()** - returns the index of the newest element in the buffer*/
    HistoryBuffer.prototype.lastIndex = function () {
        return this.hb.lastIndex();
    };

    /** **get(n)** - returns the nth element in the buffer*/
    HistoryBuffer.prototype.get = function (index) {
        return this.hb.get(index);
    };

    /** **appendArray(arr)** - appends an array of elements to the buffer*/
    HistoryBuffer.prototype.appendArray = function (arr) {
        this.hb.appendArray(arr);
    };

    /** **toArray()** - returns the content of the history buffer as an array */
    HistoryBuffer.prototype.toArray = function () {
        return this.hb.toArray();
    };

    /** **toDataSeries()** - returns the content of the history buffer into a
    flot data series*/
    HistoryBuffer.prototype.toDataSeries = function (index) {
        return this.hb.toDataSeries(index);
    };

    HistoryBuffer.prototype.registerOnChange = function (key, f) {
        this.hb.registerOnChange(key, f);
    };

    HistoryBuffer.prototype.deregisterOnChange = function (key, f) {
        this.hb.deregisterOnChange(key, f);
    };

    /** **query(start, end, step, index)** - decimates the data set at the
    provided *index*, starting at the start sample, ending at the end sample
    with the provided step */
    HistoryBuffer.prototype.query = function (start, end, step, index) {
        return this.hb.query(start, end, step, index);
    };

    /** **rangeX(index)** - returns the range of the data in the buffer */
    HistoryBuffer.prototype.rangeX = function (index) {
        return this.hb.rangeX(index);
    };

    /** **rangeY(start, end, index)** - returns the range of the data
    in a given interval of the buffer*/
    HistoryBuffer.prototype.rangeY = function (start, end, index) {
        return this.hb.rangeY(start, end, index);
    };

    /** **toJSON()** - returns the JSON of the data in the buffer */
    HistoryBuffer.prototype.toJSON = function () {
        return this.hb.toJSON();
    };

    Object.defineProperty(HistoryBuffer.prototype, 'width', {
        get: function() {
            return this.hb.width;
        }
    });

    Object.defineProperty(HistoryBuffer.prototype, 'count', {
        get: function() {
            return this.hb.count;
        },
        set: function(count) {
            this.hb.count = count;
        },
    });

    Object.defineProperty(HistoryBuffer.prototype, 'capacity', {
        get: function() {
            return this.hb.capacity;
        }
    });

    if (typeof module === 'object' && module.exports) {
        module.exports = HistoryBuffer;
    } else {
        global.HistoryBuffer = HistoryBuffer;
    }
})(window);

/* Flot plugin that makes charting easier and more efficient.

Copyright (c) 2007-2015 National Instruments
Licensed under the MIT license.
*/
/*global jQuery, requestAnimationFrame*/

(function ($) {
    'use strict';

    // flot hook which decimates the data from the historyBuffer and converts it into a format that flot understands
    function processRawData(plot, dataSeries, x, datapoints) {
        var indexMap; // this a "dictionary" from 0 based indexes in the history buffer to target values
        if (dataSeries.historyBuffer) {
            var historyBuffer = dataSeries.historyBuffer;
            indexMap = historyBuffer.indexMap;
            var data;

            var index = plot.getData().indexOf(dataSeries);

            if (index < historyBuffer.width) {
                data = dataRange(historyBuffer, index);
                dataSeries.index = index;
            } else {
                data = [];
            }

            var points = datapoints.points;

            for (var i = 0; i < data.length; i+=2) {
                points[i] = indexMap ? indexMap[data[i]] : data[i];
                points[i + 1] = data[i + 1];
            }

            points.length = data.length;
            datapoints.pointsize = 2;
            dataSeries.decimate = decimateChartData;
            dataSeries.decimatePoints = decimateChartData;
        }
    }

    function dataRange(historyBuffer, index) {
        var data = historyBuffer.rangeX(index);
        var result = [];

        if (data.xmin === undefined || data.xmax === undefined || data.deltamin === undefined) {
            return [];
        }

        result[0] = data.xmin;
        result[2] = data.xmax;
        result[4] = data.xmax - data.deltamin;
        result[1] = 0;
        result[3] = 1;
        result[5] = 2;


        return result;
    }

    function decimateChartData (series, start, end, width) {
        var historyBuffer = series.historyBuffer,
            size = end - start,
            indexMap = historyBuffer.indexMap,
            datapoints = series.datapoints,
            step = Math.floor(size / width),
            data;

        var index = series.index;
        if (index < historyBuffer.width) {
            var data = series.historyBuffer.query(start, end, step, index);
        } else {
            data = [];
        }

        var points = datapoints.points;
        for (var i = 0; i < data.length; i+=2) {
            points[i] = indexMap ? indexMap[data[i]] : data[i];
            points[i+1] = data[i+1];
        }

        points.length = data.length;
        datapoints.pointsize = 2;

        return points;
    }

    // remove old data series and trigger the computation of a new one from the history buffer
    function updateSeries(plot, historyBuffer) {
        var dataSeries = plot.getData();
        for (var i = 0; i < historyBuffer.width; i++) {
            if (typeof dataSeries[i] === 'object') {
                dataSeries[i].data = [];
                // although it would be nice to reuse data points, flot does nasty
                // things with the dataSeries (deep copy, showing with ~50% percent
                // on the timeline)
                dataSeries[i].datapoints = undefined;
            } else {
                dataSeries[i] = {
                    data: [],
                };
            }
        }

        plot.setData(dataSeries);
    }

    //readjust xmin to account for large deltamin
    function adjustDataRange(plot, s, range) {
        if (s.datapoints.points) {
            range.xmin = s.datapoints.points[0];
        }
    }

    function setYAxisRange(plot, yaxis) {
        if (yaxis.direction !== 'y' || yaxis.options.autoScale === "none")
            return;
        var i, j, k, points, pointsLength, xmin, xmax, range, index,
            dataSeries = plot.getData(),
            yseries = dataSeries
                .filter(function(series) {
                    return series.yaxis === yaxis;
                }),
            yseriesLength = yseries.length;
        for (j = 0; j < yseriesLength; j++) {
            index = yseries[j].index;
            xmin =  yseries[j].xaxis.min ? yseries[j].xaxis.min : yseries[j].xaxis.options.min;
            xmax =  yseries[j].xaxis.max ? yseries[j].xaxis.max : yseries[j].xaxis.options.max;
            if (j < yseries[j].historyBuffer.width) {
                range =  yseries[j].historyBuffer.rangeY(xmin, xmax, index);
                if (j === 0) {
                    yaxis.datamin = range.ymin;
                    yaxis.datamax = range.ymax;
                } else {
                    yaxis.datamin = Math.min(yaxis.datamin, range.ymin) || undefined;
                    yaxis.datamax = Math.max(yaxis.datamax, range.ymax) || undefined;
                }
            }
        }
    }

    // draw the chart
    function drawChart(plot) {
        plot.setupGrid(true);
        plot.draw();
    }

    // plugin entry point
    function init(plot) {
        var isShutdown = false;
        const keyForOnChangeHistoryBuffer = 'setDataCallback';
        // called on every history buffer change.
        function triggerDataUpdate(plot, historyBuffer) {
            if (!plot.dataUpdateTriggered) {
                plot.dataUpdateTriggered = requestAnimationFrame(function () { // throttle charting computation/drawing to the browser frame rate
                    if (!isShutdown) {
                        updateSeries(plot, historyBuffer);
                        drawChart(plot);
                    }
                    plot.dataUpdateTriggered = null;
                });
            }
        }

        plot.hooks.processOptions.push(function (plot) {
            var historyBuffer = plot.getOptions().series.historyBuffer; // looks for the historyBuffer option
            if (historyBuffer) {
                plot.hooks.processRawData.push(processRawData); // enable charting plugin for this flot chart
                historyBuffer.registerOnChange(keyForOnChangeHistoryBuffer, function () {
                    triggerDataUpdate(plot, historyBuffer); // call triggerDataUpdate on every historyBuffer modification
                });
                updateSeries(plot, historyBuffer);

                plot.hooks.adjustSeriesDataRange.push(adjustDataRange);

                plot.hooks.setRange.push(setYAxisRange);

                plot.hooks.shutdown.push(function () {
                    historyBuffer.deregisterOnChange(keyForOnChangeHistoryBuffer);
                    isShutdown = true;
                });
            }
        });
    }

    $.plot.plugins.push({
        init: init,
        name: 'charting',
        version: '0.3'
    });
})(jQuery);
